godel_surface_detection/src/scan/profilimeter_scan.cpp

```diff
@@ -0,0 +1,202 @@
+      x += dx;
+      y += dy;
+    }
+
+    return slices;
+  }
+
+  std::vector<PolygonPt> interpolateAlongAxis(const RotatedRect& rect, double ds)
+  {
+    std::vector<PolygonPt> pts;
+    double dx, dy; // 
+    float x, y;
+    int n;
+    n = rect.w / ds;
+    dx = std::cos(rect.a);
+    dy = std::sin(rect.a);
+    x = rect.x - dx * rect.w/2.0;
+    y = rect.y - dy * rect.w/2.0;
+
+    // transform unit vector
+    dx *= ds; 
+    dy *= ds;
+
+    for (int i = 0; i < n; ++i)
+    {
+      PolygonPt pt;
+      pt.x = x;
+      pt.y = y;
+      x += dx;
+      y += dy;
+      pts.push_back(pt);
+    }
+
+    return pts;
+  }
+
+  std::vector<PolygonPt> makeStitch(const PolygonPt& a, const PolygonPt& b, double ds)
+  {
+    double dx = b.x - a.x;
+    double dy = b.y - a.y;
+    double s = std::sqrt(dx*dx + dy*dy);
+    // steps
+    size_t n = s / ds;
+    // unit vector
+    dx /= s;
+    dy /= s;
+
+    std::vector<PolygonPt> result;
+    for (size_t i = 0; i < n; ++i)
+    {
+      PolygonPt pt;
+      pt.x = a.x + i * dx * ds;
+      pt.y = a.y + i * dy * ds;
+      result.push_back(pt);
+    }
+    return result;
+  }
+
+  std::vector<PolygonPt> stitchAndFlatten(const std::vector<std::vector<PolygonPt> >& paths, double ds)
+  {
+    std::vector<PolygonPt> result;
+    bool forward = true;
+    for (std::size_t i = 0; i < paths.size(); ++i)
+    {
+      // copy current path in given direction
+      if (forward)
+      {
+        result.insert(result.end(), paths[i].begin(), paths[i].end());
+      }
+      else
+      {
+        result.insert(result.end(), paths[i].rbegin(), paths[i].rend());
+      }
+
+      forward = !forward;
+
+      // check to see if there is a next strip
+      if (i+1 < paths.size())
+      {
+        // if so, stitch to the next point
+        std::vector<PolygonPt> stitch = makeStitch(*result.rbegin(), (forward ? *(paths[i+1].begin()) : *(paths[i+1].rbegin())), ds);
+        result.insert(result.end(), stitch.begin(), stitch.end());
+      }
+    }
+
+    return result;
+  }
+
+
+} // end anon namespace
+
+std::vector<PolygonPt> 
+godel_surface_detection::generateProfilimeterScanPath(const PolygonBoundary& boundary,
+                                                      const godel_msgs::ScanPlanParameters &params)
+{
+  static const double DS = 0.01;
+
+  // Step 1 -> compute (oriented?) bounding box
+  RotatedRect bbox = simpleBoundingBox(boundary);
+  
+  // Step 2 -> slice bounding box into strips
+  std::vector<RotatedRect> slices = sliceBoundingBox(bbox, params.scan_width, params.overlap);
+  
+  // Step 3 -> generate set of dense points along center of each strip
+  std::vector<std::vector<PolygonPt> > slice_points;
+  slice_points.reserve(slices.size());
+  for (std::size_t i = 0; i < slices.size(); ++i)
+    slice_points.push_back(interpolateAlongAxis(slices[i], DS));
+
+  // Step 4 -> connect slices together
+  std::vector<PolygonPt> pts = stitchAndFlatten(slice_points, DS); 
+
+  return pts;
+}
```

Comment:
Are the array sizes returned by these vector generating methods rather large? If so perhaps it be more efficient to set them up as output arguments.